_posts/2017/2017-16-01-testing-async-interactions.md

```diff
@@ -0,0 +1,246 @@
+    assertThatJson(parsedJson).field("name").isEqualTo("Awesome Jeans");
+    assertThatJson(parsedJson).field("purchasePrice").isEqualTo(25.00);
+    assertThatJson(parsedJson).field("id").matches("[0-9]+");
+    assertThatJson(parsedJson).field("salesPrice").isEqualTo(49.99);
+  }
+}
+```
+
+So if we would change an attribute name in `Product` we would have a failing test.
+**This shields us from making accidental incompatible contract-relevant changes**.
+
+The contracts of the product service are published in a stub jar to a maven repository.
+
+This stub jar can be used by the checkout service.
+This is done by stating the stub information in the [application configuration](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/main/resources/application.yml) of the checkout service.
+
+```yml
+stubrunner:
+  work-offline: true
+  ids: 'com.epages:product-service'
+```
+
+Now all we need to do is annotate our improved test on checkout side with `@AutoConfigureStubRunner` and use the `StubTrigger` to send the stub message in.
+The `spring-cloud-contract-stub-runner` downloads the stub jar with our contract from a maven repository and emits it into our application.
+This is done without interaction with a running broker.
+After the event is emitted we can run our assertions to make sure the correct effect can be observed.
+In our case, we assert that the correct product reference has been saved.
+Our [refactored test](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/test/java/com/epages/checkout/ProductSubscriberContractTest.java) looks like this:
+
+```java
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = CheckoutServiceApplication.class)
+@AutoConfigureStubRunner
+public class ProductSubscriberContractTest {
+
+  @Autowired
+  private StubTrigger stubTrigger;
+
+  @Autowired
+  private ProductRefRepository productRefRepository;
+
+  @Test
+  public void should_handle_product_created_event() {
+    //emit an event from contract with label "product.created.event"
+    stubTrigger.trigger("product.created.event");
+
+    then(productRefRepository.exists(9L)).isTrue();
+  }
+}
+```
+
+Basically, the test is covering the same aspects as our first version.
+But **it always picks up the latest contract definitions**.
+This can be called a provider contract test approach. The owner of the interface publishes the information about the structure of the event it emits (e.g. the contract).
+A consumer can use the contract to assert that events that follow the contract can be processed.
+
+## Benefits and Downsides
+
+The main benefit of this approach is the ability to always test against the contract published by the provider.
+Tests in the consumer of a message break if it no longer is able to process messages that follow the contract.
```

Comment:
Tests in the consumer of a message break if it is no longer able to process messages that follow the contract.